deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java

```diff
@@ -259,6 +249,53 @@ boolean canDeploy(EnvironBean envBean, String host, AgentBean agentBean, String
         }
     }
 
+    boolean canDeployWithConstraint(String hostId, EnvironBean envBean) throws Exception {
+        String envId = envBean.getEnv_id();
+        String constraintId = envBean.getDeploy_constraint_id();
+
+        try {
+            LOG.info("DeployWithConstraint env {}: verify active agents for host {} constraint {}", envId, hostId, constraintId);
+
+            DeployConstraintBean deployConstraintBean = deployConstraintDAO.getById(constraintId);
+            String tagName = deployConstraintBean.getConstraint_key();
+            HostTagBean hostTagBean = hostTagDAO.get(hostId, tagName);
+            if (deployConstraintBean.getState() != TagSyncState.FINISHED) {
```

Comment:
I think canDeployWithConstraint function should not care or know about the sync process at all. simply look at the host tags:
if no such tag or tag value is undefined, return false
otherwise, check the tag value.

Reply:
this also implies, deploy does not have to wait for sync explicitly, both deploy and sync could be on going at the same time, concurrently

Reply:
make sense, just removed the update tag sync state logic from pingHandler, which should only happen in tag sync worker